### PR TITLE
Add origin to Transformation

### PR DESCRIPTION
**Pull request by ffreyer**
# Description

Related: #4445, #2637

This adds an `origin` to `Transformation` so that `rotate!()` and `scale!()` can be relative to something other than `Vec3(0)`. It can be set with `translate_origin!()` just like `translate!()`.

Example:

```julia
p = rand(Point2f)
rect = Rect2f(p .- 0.1, Vec2f(0.2))

fig = Figure()
sl = Slider(fig[1, 1], range = range(0, 2pi, length=101))
ax = Axis(fig[2, 1], aspect = DataAspect())
p1 = poly!(ax, rect, color = :orange)
p2 = lines!(ax, rect, color = :red)
xlims!(ax, -1.5, 1.5)
ylims!(ax, -1.1, 1.1)

on(sl.value) do angle
    center = to_ndim(Vec3f, p, 0)

    R = Makie.rotationmatrix_z(angle)
    T = Makie.translationmatrix(-center)
    Tinv = Makie.translationmatrix(center)
    p1.transformation.model[] = Tinv * R * T # Did we break setting plot.model?
    
    translate_origin!(p2, p)
    Makie.rotate!(p2, angle)
end

fig
```

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] Added an entry in CHANGELOG.md (for new features and breaking changes)
- [x] Added or changed relevant sections in the documentation
- [x] Added unit tests for new algorithms, conversion methods, etc.
- [ ] Added reference image tests for new plotting functions, recipes, visual options, etc.

### CODE REVIEW

File: docs/src/reference/generic/transformations.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+You can set the transformation function of a plot by updating `plot.transformation.transform_func[] = new_func`.
+This will also change the transformation function of each child plot.
+Note that this will be reset when the plots parent transformation function changes, e.g. if `ax.xscale` is set in the example above.
+Another option is to create the plot with an explicitly given `Transformation` object, which can detach the plots transformation from its parent.
+
+## Scene Transformation
+
+The Scene also holds onto a `Transformation` object.
+It's `transform_func` acts as a default for any plot added to the scene.
+Its `model` transformation acts as a secondary transformation to any plot, i.e. it is applied aftr the plots own model transformation `scene.transformation.model[] * plot.transformation.model[]`.
```
Comment (bjarthur) [inline docs/src/reference/generic/transformations.md:63]: after not aftr

File: docs/src/reference/generic/transformations.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+For example, you may want to apply `transform_func` before running a triangulation algorithm in a recipe so that the triangulation doesn't get distorted.
+Or you may want to chain transformations of a series of plots to keep them relative to each other (see below).
+There are a few constructors which are helpful to know for this.
+
+If you want to fully detach a plot from its parents transformations, you can create it with `transformation = Transformation()`.
+If you want to remove only the `transform_func` but not model transformations, you can use `transformation = Transformation(parent, transform_func = identity)`.
+You can also pass different starting values for `translation`, `scale` and `rotation` to these functions.
+This will not affect whether the parents model transformations are considered.
+
+As an example, here is a two arms on a cart raising a box with a rope.
```
Comment (bjarthur) [inline docs/src/reference/generic/transformations.md:78]: "here are two..." ?

**Comment by bjarthur** [comment]
looking forward to this!

one other suggestion in addition to the typos:  i find the name `translate_origin!` a little confusing as the plot isn't actually getting translated.  how about `set_origin!` instead?  or maybe just `origin!`?  or, maybe `rotate!` and `scale!` could take an `origin` kwarg?

**Comment by jkrumbiegel** [comment]
> maybe rotate! and scale! could take an origin kwarg?

I would prefer that I think, or maybe `around` so it's `rotate!(x, ..., around = p)`? Or `relative_to`

**Comment by ffreyer** [comment]
Having origin as an argument in rotate!() and scale!() was also my first thought but I think with how transformations work it's better to pull it out. Transformations don't consider the order you call things in. You're effectively setting or changing the state with `translate!()`, `rotate!()` etc. To me `rotate!(..., origin = ...)` makes it seem like you're setting the origin for this specific call, while you would actually set it for the transformation as a whole. I think a separate function for origin reflects that better.

I'd be fine with renaming the function. I also had `set_origin!()` originally but switched when I implemented the `Accum/Absolute` interface since `set` sounds like just overwriting (no `origin = origin + arg` which Accum does)

**Pull request merged**